File: codingTest/2week-assignment/day6/B20055.py
# https://www.acmicpc.net/problem/20055
# 컨베이어벨트


N, K = map(int, input().split())  # N과 K 값을 입력받음
belt = list(map(int, input().split()))  # 컨베이어 벨트의 내구도를 입력받아 리스트에 저장
robot = [False] * N  # 로봇의 위치를 나타내는 리스트 초기화

level = 0  # 현재 레벨(단계)을 나타내는 변수 초기화

while True: # 반복문 시작
    level += 1
    tmp = belt[-1]
    for i in range(N*2-1, 0, -1): # 벨트를 한 칸씩 회전시킴
        belt[i] = belt[i-1]
    belt[0] = tmp
    print("belt:", belt)  # 현재 벨트의 상태 출력

    for i in range(N-1, 0, -1):  # 로봇을 한 칸씩 이동시킴
        robot[i] = robot[i-1]
    robot[0] = False  # 첫 번째 위치의 로봇을 False로 설정

    robot[N-1] = False  # 마지막 위치의 로봇을 내림

    for i in range(N-1, 0, -1):  # 로봇이 이동할 수 있는지 확인하고 이동시킴
        if robot[i-1] and not robot[i] and belt[i] > 0:
            robot[i-1] = False
            robot[i] = True
            belt[i] -= 1  # 이동한 위치의 내구도를 1 감소


    robot[N-1] = False

    if belt[0] > 0:
        robot[0] = True
        belt[0] -= 1

    cnt = sum(1 for x in belt if x == 0)
    if cnt >= K:
        break

print(level)

File: java-version/src/week2/day5/B20055.java
package week2.day5;
//https://www.acmicpc.net/problem/20055
//컨베이어 벨트위의 로봇


//우선 한칸 이동  이동할때 로봇이 있으면 그로봇도 한칸이동
//로봇이 마지막 지점에 있으면 로봇 제거

// 가장 먼저 올라간 로봇 순서대로 한칸 더 이동할수있으면 이동
//이동후 수명1감소
//로봇이 마지막지점에 온다면 제거

//시작지점에 로봇을 올릴수있으면 올려라

// 내구도가 0개가 k개 이상이면 종료

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class B20055 {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());

        int[] belt = new int[N*2];
        boolean[] robot = new boolean[N];

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N*2; i++) {
            belt[i] = Integer.parseInt(st.nextToken());     // 컨베이어 벨트의 내구도를 입력받아 배열에 저장합니다.

        }

        int level = 0;     // 진행 단계를 나타내는 변수입니다.


        while(true) {
            level++;          // 레벨을 1 증가시킵니다.
            // 1단계
            int tmp = belt[N*2-1];
            for (int i = N*2-1; i > 0; i--) {
                belt[i] = belt[i-1];
            }         // 로봇도 한 칸 이동시킵니다.

            belt[0] = tmp;

            for (int i = N-1; i > 0; i--) {
                robot[i] = robot[i-1];
            }

            robot[0] = false;
            robot[N-1] = false;

            // 2단계
            for (int i = N-1; i > 0; i--) {
                if(robot[i-1] && !robot[i] && belt[i]>0) {
                    robot[i-1] = false;
                    robot[i] = true;
                    belt[i]--;
                    robot[N-1] = false;
                }
            }

            // 3단계
            if(belt[0]>0) {
                robot[0] = true;
                belt[0]--;
            }

            // 4단계
            int cnt = 0;
            for (int i = 0; i < N*2; i++) {
                if(belt[i] == 0) cnt++;
            }
            if(cnt>=K) break;

        }

        System.out.println(level);
    }
}
